### PR TITLE
Steensgaard implementation and adding ctx sensitivity at call sites

### PR DESCRIPTION
**Pull request by yousifpatti**
- [x]  Steensgaard base class
- [x]  Stack support
- [x]  Globals support
<del>- [ ]  Heap Support</del> In the coming iteration
- [x]  <del>Context Sensitivity</del> Set based SSA

### CODE REVIEW

**Review by ailrst — COMMENTED** [review]
It doesn't look like the `ssa_id` is doing the right thing, so the steensgaard and constprop with ssa probably don't work correctly but the ANR and RNA are probably good, although it would be good to have some basic tests.

File: src/main/scala/analysis/SSAForm.scala [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    varMaxTracker(varName) = ret + 1
+    ret
+
+  def transformVariables(vars: Set[Variable], block: Block, proc: Procedure): Unit = {
+    vars.foreach(v => {
+      v.ssa_id = blockBasedMappings.getOrElseUpdate(
+        (block, v.name),
+        context.getOrElseUpdate((proc, v.name), mutable.Set(getMax(v.name)))
+      )
+    })
```
[the remark this answers is not in the document]
Comment (ailrst) [reply src/main/scala/analysis/SSAForm.scala:28]: On second look, this is just creating new SSA IDs without caring about the specific definition? In that case it is ok?

File: src/main/scala/ir/Expr.scala [hunk 2]
```diff
@@ -1,8 +1,10 @@
 package ir
 
 import boogie._
+import scala.collection.mutable
 
 sealed trait Expr {
+  var ssa_id: mutable.Set[Int] = mutable.Set[Int]()
```
[the remark this answers is not in the document]
Comment (yousifpatti) [reply src/main/scala/ir/Expr.scala:7]: I will have to add this as a field for the Variable trait as Register and LocalVar are immutable case classes. Would this work?

File: src/main/scala/analysis/SSAForm.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (yousifpatti) [reply src/main/scala/analysis/SSAForm.scala:28]: This is meant to create a new SSA ID without relying on any particular order if the variable (register) has not been defined before. If the variable has been defined before in that procedure, then we want to assume the last definition applies. The caller of `transformVariables` is responsible for controlling the order of processing statements.

**Comment by yousifpatti** [comment]
I am currently writing tests and it will need to be moved to use the IR iterator at some point (working on it in another branch). The Points-to constraints may need to be refined as they are not precisely Steensgaard constraints because they were not meant for Binary-level pointer analysis. The current constraints are a highly modified version of the Steensgaard constraints and work is needed to verify soundness.
The current progress can be merged through this PR while those aspects are being worked on. I will just get it up to date with main first and resubmit PR review request.

**Comment by l-kent** [comment]
Ah, what I'm really interested in everything being moved over to the new iterator, and I was wanting to start work on that, but it's difficult to do so when there's significant outstanding changes to the existing analyses

The simplest thing would be to merge this branch into staging once the current merge happening there is finished, and then we can merge all of staging into main. It's not an issue if the points-to and Steengaard analyses aren't perfect yet, just as long as nothing else with the existing analyses has regressed.

**Comment by l-kent** [comment]
Is this ready to merge into staging now?

**Comment by yousifpatti** [comment]
@l-kent points-to and analyses changes has been merged into staging

**Pull request merged**